### ISSUE TITLE
False Negative: Assignments inside logical expressions are missed in non-canonical forms.

### ISSUE THREAD

**Issue by Carlson-JLQ** [google/error-prone#5639]
# False Negative: Assignments inside logical expressions are missed in non-canonical forms.

Version
2.48.0

## Checker
- Checker id: `LogicalAssignment`
- Checker description: This checker detects assignments used as boolean conditions in control flow statements (if, while, do-while, for) where a boolean expression is expected.

## Description of the false negative
`diff/output` marks 1 false negatives for `LogicalAssignment`. Across these samples, the benchmark is still exercising the same underlying bug pattern, but the implementation appears to lose track of it once the source deviates from one canonical shape. The misses read more like matching gaps than deliberate exclusions.

### `PosCase2.java`
Source: `output/codeql/NoAssignInBooleanExprs/src/main/java/scensct/core/pos/PosCase2.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Assignment inside non-comparison binary expression used as condition should be flagged as suspicious.

```java
// Assignment inside non-comparison binary expression used as condition should be flagged as suspicious.
package scensct.core.pos;

public class PosCase2 {
    public static void main(String[] args) {
        int a = 0;
        int b = 0;
        // Assignment inside arithmetic expression used as condition
        if ((a = 5) + b > 0) { // [REPORTED LINE]
            System.out.println("test");
        }
    }
}
```

## Cause analysis
The rule is documented as follows: This checker detects assignments used as boolean conditions in control flow statements (if, while, do-while, for) where a boolean expression is expected.

 The missed cases suggest that the implementation is narrower than the rule description and too dependent on one preferred AST shape.

Once the same behavior is expressed through a helper, an overload, a modifier such as `abstract` or `native`, a different comment position, or a slightly rearranged control-flow structure, the report disappears even though the benchmark still treats the case as in-scope.

For `LogicalAssignment`, the practical improvement would be to generalize the match so it follows the underlying behavior rather than one exact spelling of it.

## References
None known.
